src/storage/src/hummock/utils.rs

```diff
@@ -121,3 +124,96 @@ where
     })
     .saturating_sub(1) // considering the boundary of 0
 }
+
+struct MemoryLimiterInner {
+    total_size: AtomicU64,
+    notify: Notify,
+    check_count: AtomicUsize,
+    quota: u64,
+}
+
+impl MemoryLimiterInner {
+    pub fn may_notify(&self, quota: u64) {
+        while self.check_count.load(AtomicOrdering::SeqCst) > 0 {}
+        let total_size = self.total_size.fetch_sub(quota, AtomicOrdering::Release);
+        if total_size >= self.quota {
+            self.notify.notify_waiters();
+        }
+    }
+}
+
+pub struct MemoryLimiter {
+    inner: Arc<MemoryLimiterInner>,
+}
+
+pub struct MemoryTracker {
+    limiter: Arc<MemoryLimiterInner>,
+    quota: u64,
+}
+
+use std::sync::atomic::Ordering as AtomicOrdering;
+
+impl MemoryLimiter {
+    pub fn new(quota: u64) -> Self {
+        Self {
+            inner: Arc::new(MemoryLimiterInner {
+                total_size: AtomicU64::new(0),
+                notify: Notify::new(),
+                check_count: AtomicUsize::new(0),
+                quota,
+            }),
+        }
+    }
+
+    pub fn try_require_memory(&self, quota: u64) -> bool {
+        if quota > self.inner.quota {
+            return false;
+        }
+        self.inner.total_size.load(AtomicOrdering::Acquire) + quota < self.inner.quota
+    }
+
+    pub async fn require_memory(&self, quota: u64) -> Option<MemoryTracker> {
+        if quota > self.inner.quota {
```

Comment:
The current logic seems complex and involve spin-loop during release, which may become a waste of compute resource or even deadlock.

We may use CAS to simplify the code
```
async fn acquire(&self, quota: u64) {
    let current_size = self.inner.total_size.load(SeqCst);
    if current_size + quota < self.inner.quota {
        if self.inner.total_size.compare_exchange(current_size, current_size + quota, SeqCst, SeqCst).is_ok() {
            return;
        }
    }

    loop {
        let notified = self.inner.notify.notified();
        let current = self.inner.total_size.load(SeqCst);
        if current + quota < self.inner.quota && 
            self.inner.total_size.compare_exchange(current, current + quota, SeqCst, SeqCst).is_ok() {
            return;
        }
        notified.await;
    }
}

fn release(&self, quota: u64) {
    self.inner.total_size.fetch_sub(quota, SeqCst);
    self.inner.notify.notify_waiters();
}
```

The above code is the general idea. Memory order can be further optimized.

Reply:
`compare_exchange` is more complex than `fetch_add`....
But I found that `wait_count` is not necessary, I would remove it.